basic_pro branch code update.

maynardGK opened the pull request:
These are changes to basic_fun to cleanup the code and to
 advance GDL functionality.

  OBJ_DESTROY is updated and its previous shortcomings should be gone.

  bug 708 indicated an interpreter error when certain conditions came up.
  Now, more robust checks are made that avoids coding with try ... catch exception throwing.
  test_bug_708.pro is provided  to verify healthy results for at least this situation.

  In HEAP_GC, if there are no heap values remaining in either the pointer or the object heaps,  GDLInterpreter::ResetHeap() is called, which will also reset the numbering of the pointer and object indices.

  Other modifications to basic_pro are pending in other pull requests:
The delvar-rnew branch brings in the DELVAR main-level command
The Windows SPAWN command is improved with the code provided in the file_etc PR.

Greg Jung (maynardGK)
gvjung@gmail.com

maynardGK committed:
basic_pro branch code update.

  These are changes to basic_fun to cleanup the code and to
 advance GDL functionality.

  OBJ_DESTROY is updated and its previous shortcomings should be gone.

  bug 708 indicated a bad interpreter error when certain conditions came up
  Alternative code to fix this includes possibly throwingf an exception, this code
  does not accept that possibility and avoids coding with try ... catch.
  test_bug_708.pro is provided  to verify that for at least this situation, no
  exception throwing is needed.

  In HEAP_GC, if there are no heap values remaing in either the pointer or the object heaps,
 a GDLInterpreter::ResetHeap() is called which will also
 reset the numbering of the pointer and object indices.

GillesDuvert committed:
Update test_obj_destroy.pro

GillesDuvert commented:
@maynardGK , thanks!
Please check, test_obj_destroy does not pass.
--> obj_destroy cannot have more than 1 object in the list.
--> tested by hand, crashes on "obj_destroy,llist " on ObjCleanup, line 346 in basic_pro.cpp

maynardGK committed:
Refurb of launch_cmd (called from spawn_pro fror Windows)

  This modification takes the (c char) -> (w_char) manipulations into the launch_cmd
routine, so that the calling routine, spawn_pro for WIN32 case, only sees UTF8
strings. MultiByteToWideChar only appears prior to its need in launch_cmd, which invokes
CreateProcessW. This only affects Windows operation.
  Also revised to use the current directory for the spawned process.

Greg Jung (aka maynardGK)
gvjung@gmail.com

slayoo commented:
Closing and reopening to trigger codecov and appveyor runs

GillesDuvert committed:
obj_destroy cannot/must not destroy 2 objects simultaneously.

maynardGK committed:
Merge branch 'basic_pro' of https://github.com/maynardGK/gdl into basic_pro

maynardGK committed:
OBJ_DESTROY: Correct parameter list interpretation.

  When obj_destroy is called with more than a single parameter, the extra parameters are to
be passed on to the object's ::cleanup procedure.  This is tested in a rudimentary
fashion in test_obj_destroy.  The built-in OBJ_DESTROY will ignore any parameters but the first
(thereby passing them, via e->ObjCleanup(), to the compiled ::CLEANUP method. Thanks to Gilles
 for pointing this out.

maynardGK committed:
Whoops.

maynardGK committed:
Whoops again!

maynardGK commented:
will close this now that I have merged the inter-related branches into branch "combo" which built
and produced the following result on Linux "make check:"


7% tests passed, 7 tests failed out of 211

Total Test time (real) = 140.14 sec

The following tests FAILED:
         94 - test_fft_leak.pro (Failed)
         97 - test_file_delete.pro (Failed)
        142 - test_n_tags.pro (Failed)
        152 - test_parse_url.pro (Failed)
        175 - test_resolve_routine.pro (Failed)
        177 - test_rounding.pro (Failed)
        199 - test_total.pro (Failed)

 combo = legacy_list + basic_fun + basic_pro + gdlhelp_fin + legacy_hash
    
      This brings the list/hash code to a level just prior to the final enhancement that is ready
    to go, which allows in-place extraction and insertions from lists and hashes.
    file_etc branch has not yet been merged, it is not dependent on this.
    
      Changes in GDLParser, GDLInterpreter, etc. should all have their corresponding code in gdlc.g
    and gdlc.i.g.
    
    Greg Jung 29 May 2018
    gvjung@gmail.com
